### ISSUE TITLE
pbsdsh is not always used as default bootstrap for recent Intel MPI

### ISSUE THREAD

**Issue by boegel** [hpcugent/vsc-mympirun#114]
The way in which an `MPI` subclass is selected currently requires that only only of the MPI subclasses matches (because the subclasses of `MPI` are processed in no particular order).

This is no longer the case since #110 since the version range is non-exclusive between `IntelHydraMPI` and the new `IntelHydraMPIPbsdsh` classes.

This can be trivially fixed with the following patch:

```diff
diff --git a/lib/vsc/mympirun/mpi/intelmpi.py b/lib/vsc/mympirun/mpi/intelmpi.py
index f194e7d..faed80d 100644
--- a/lib/vsc/mympirun/mpi/intelmpi.py
+++ b/lib/vsc/mympirun/mpi/intelmpi.py
@@ -205,7 +205,7 @@ class IntelHydraMPI(IntelMPI):
 
     _mpiscriptname_for = ['ihmpirun']
 
-    _mpirun_version = staticmethod(lambda x: LooseVersion(x) >= LooseVersion("4.1.0.0"))
+    _mpirun_version = staticmethod(lambda x: LooseVersion(x) >= LooseVersion('4.1.0.0') and LooseVersion(x) < LooseVersion('5.0.3'))
 
     HYDRA = True
     HYDRA_LAUNCHER_NAME = "bootstrap"
```

It would be nicer to provide a helper function in `mpi.py` to make specifying the version range easier, something like:

```python
def version_in_range(version, lower_limit, upper_limit):
    """
    Check whether version is in specified range

    :param lower_limit: lower limit for version (inclusive), no lower limit if None
    :param upper_limit: upper limit for version (exclusive), no upper limit if None
    """
    in_range = True
    if lower_limit is not None and LooseVersion(version) < LooseVersion(lower_limit):
        in_range = False
    if upper_limit is not None and LooseVersion(version) >= LooseVersion(upper_limit):
        in_range = False
    return in_range
```

```python
from vsc.mympirun.mpi.mpi import version_in_range
...
class IntelHydraMPI(IntelMPI):
    _mpirun_version = staticmethod(lambda ver: version_in_range(ver, '4.1.0.0', '5.0.3'))
...
class IntelHydraMPIPbsdsh(IntelHydraMPI):
    _mpirun_version = staticmethod(lambda ver: version_in_range(ver, '5.0.3', None))
```

**Comment by boegel** [comment]
This doesn't come forward in the `end2end` tests because they explicitly specify `--setmpi ihmpirun`, in which case (by accident) `IntelHydraMPIPbsdsh` gets considered before `IntelHydraMPI` for some reason...

@Caylo: What would be needed to avoid having to hard specify `ihmiprun` to make the tests run as if Intel MPI is being used? Can we also control the version, to make sure we get different behaviour for Intel MPI >= v5.0.3 and < v5.0.3?

**Comment by boegel** [comment]
Worth mentioning: using `pbsdsh` shaves off quite a bit of startup time (this is with 60 cores on golett, spread across 3 nodes):

```
$ time mympirun --launcher ssh ./mpi_hello > /dev/null

real	0m4.849s
user	0m2.562s
sys	0m3.262s
$ time mympirun --launcher pbsdsh ./mpi_hello > /dev/null

real	0m3.627s
user	0m1.826s
sys	0m3.563s
```

**Comment by boegel** [comment]
fixed in #115
